packages/malloc/src/pool.ts

```diff
@@ -307,3 +372,67 @@ export class MemPool implements IMemPool {
         block.next = ptr;
     }
 }
+
+
+const MEM_BLOCK_SIZE_OFFSET = 0;
+const MEM_BLOCK_NEXT_OFFSET = MEM_BLOCK_SIZE_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+const MEM_BLOCK_DATA_OFFSET = MEM_BLOCK_NEXT_OFFSET + Uint32Array.BYTES_PER_ELEMENT;
+
+/**
+ * All of the addresses that saved are the memory address, and not the data address.
+ * The user see the data address, we need to convert them in and out
+ */
+class MemBlockWrapper implements MemBlock {
+    public static OVERHEAD = MEM_BLOCK_DATA_OFFSET;
+
+    public static toDataAddress(blockAddress: number) {
+        return blockAddress + MEM_BLOCK_DATA_OFFSET;
+    }
+
+    public static toMemoryAddress(dataAddress: number) {
+        return dataAddress - MEM_BLOCK_DATA_OFFSET;
+    }
+
+    constructor(
+        private dataView: DataView, 
+        /**
+         * This is the memory address, not the data address
+         */
+        public readonly addr: number) {}
+
+    public get next(): MemBlock | null {
+        if (this.nextAddress !== 0) {
+            return new MemBlockWrapper(this.dataView, this.nextAddress);
+        }
+
+        return null;
+    }
+
+    public set next(value: MemBlock | null) {
+        this.nextAddress = value ? value.addr : 0;
+    }
+
+    public get size() {
+        return this.dataView.getUint32(this.addr + MEM_BLOCK_SIZE_OFFSET)
```

Comment:
Just to illustrate approach described above (i.e. using a u32 view of entire buffer). Since alignment is currently always 8 this will be safe to do (actually, we could make alignment configurable to multiples of 4, e.g. for SIMD we'll soon need 16 byte alignment):

```ts
return this.u32[(this.addr >> 2) + MEM_BLOCK_SIZE];
```